src/Host-SDK/Temporal.Worker.Workflows.Dynamic/DynamicWorkflowsApi.cs

```diff
@@ -0,0 +1,295 @@
+    /// it if and when a matching signal handler is configures", "ignore the signal" or other
+    /// (<see cref="DefaultSignalHandlerPolicy"/>).
+    /// If SignalName is not matched by the default policy's matcherRegex, the policy does not apply.
+    /// In that case, the signal will fall through to the hardcoded fafault provided by
+    /// <see cref="BasicWorkflowBase.HandleSignalAsync(String, PayloadsCollection, WorkflowContext)"/>.
+    /// This will likely be ignoring the signal logging an error.</summary>
+    public interface IDynamicWorkflowController
+    {
+        IHandlerCollection<Func<string, IDataValue, DynamicWorkflowContext, Task>> SignalHandlers { get; }
+        IHandlerCollection<Func<string, IDataValue, DynamicWorkflowContext, IDataValue>> QueryHandlers { get; }
+
+        SignalHandlingOrderPolicy SignalHandlingOrderPolicy { get; set; }
+
+        SignalHandlerDefaultPolicy SignalHandlerDefaultPolicy { get; set; }
+        QueryHandlerDefaultPolicy QueryHandlerDefaultPolicy { get; set; }
+    }
+
+    public interface IHandlerCollection<THandler> where THandler : class
+    {
+        int Count { get; }
+        void Clear() { }
+        void GetAt(int index, out string matcherRegex, out THandler handler);
+        void RemoveAt(int index);
+        void RemoveAt(int index, out string matcherRegex, out THandler handler);
+        bool TryAdd(string matcherRegex, THandler handler);
+        bool TryInsert(int index, string matcherRegex, THandler handler);
+        bool TryUpdateAt(int index, string matcherRegex, THandler handler);
+        bool TryGetByRegexPattern(string matcherRegex, out int index);
+        bool TryGetByRegexPattern(string matcherRegex, out int index, out THandler handler);
+        bool TryFindFirstMatch(string itemToMatch, out int index);
+        bool TryFindFirstMatch(string itemToMatch, out int index, out string matcherRegex, out THandler handler);        
+    }
+
+    /// <summary>This setting affects the processing order of signals. It is independent of SignalHandlerDefaultPolicy.
+    /// However, unless <see cref="SignalHandlerDefaultPolicy" /> is caching, any of these settings result in the same
+    /// semantic behavior. Should this instead be a parameter to
+    /// <see cref="SignalHandlerDefaultPolicy.CacheAndProcessWhenHandlerIsSet(String)" />?
+    /// <br />
+    /// Need to design when a a cached signal is handled if a matching handler is added:
+    ///  - immediately?
+    ///  - workflow should call some kind of HandleNewlyEnabledSingals() API?
+    ///  - at the end of the current workflow task (at the start there was no handler yet)?
+    ///  - at the satart of next workflow task, before additional signals that arrive in the meantime?
+    ///  - something else?
+    /// This choice may simplify some of the <c>SignalHandlingOrderPolicy</c>-options.</summary>
+    public class SignalHandlingOrderPolicy
```

Comment:
I'm not convinced we need this at least not for the initial proposal.

Reply:
Agreed. Not the first ting to implement. Though, it's to have a vision for how dynamic wf APIs should look like in the mid-term.